iodata/formats/fchk.py

```diff
@@ -435,3 +439,225 @@ def _triangle_to_dense(triangle: np.ndarray) -> np.ndarray:
         result[:irow + 1, irow] = triangle[begin:end]
         begin = end
     return result
+
+
+# The fchk file has a very rigid format, to dump the information are
+# theses functions, both scalars and arrays, integer and real(float) variables
+def _dump_integer_scalars(name: str, val: int, f: TextIO):
+    """Dumper for a scalar integer."""
+    print("{0:40}   I     {1:12d}".format(name, int(val)), file=f)
+
+
+def _dump_real_scalars(name: str, val: float, f: TextIO):
+    """Dumper for a scalar float."""
+    print("{0:40}   R     {1: 16.8E}".format(name, float(val)), file=f)
+
+
+def _dump_integer_arrays(name: str, val: np.ndarray, f: TextIO):
+    """Dumper for a array of integers."""
+    nval = val.size
+    if nval != 0:
+        np.reshape(val, nval)
+        print("{0:40}   I   N={1:12}".format(name, nval), file=f)
+        k = 0
+        for i in range(nval):
+            print("{0:12}".format(int(val[i])), file=f, end='')
+            k += 1
+            if k == 6 or i == nval - 1:
+                print("", file=f)
+                k = 0
+
+
+def _dump_real_arrays(name: str, val: np.ndarray, f: TextIO):
+    """Dumper for a array of float."""
+    nval = val.size
+    if nval != 0:
+        np.reshape(val, nval)
+        print("{0:40}   R   N={1:12}".format(name, nval), file=f)
+        k = 0
+
+        for i in range(nval):
+            print("{0: 16.8E}".format(val[i]), file=f, end='')
+            k += 1
+            if k == 5 or i == nval - 1:
+                print("", file=f)
+                k = 0
+
+
+@document_dump_one(
+    "Gaussian Formatted Checkpoint",
+    ['atcoords', 'atnums', 'energy', 'mo', 'obasis'],
+    ['atcharges', 'atcorenums', 'atfrozen', 'atgradient', 'athessian', 'atmasses', 'lot',
+     'one_rdms', 'obasis_name', 'extra', 'moments'])
+def dump_one(f: TextIO, data: IOData):
+    """Do not edit this docstring. It will be overwritten."""
+    # write title
+    print("{0:72}".format(data.title or "FCHK generated by IOData"), file=f)
+
+    # write run type, level of theory, and basis set name (all in uppercase)
+    items = [getattr(data, item) or "NA" for item in ["run_type", "lot", "obasis_name"]]
+    if items[0] == "energy":
+        items[0] = "SP"
+    print(f"{items[0].upper():10s}{items[1].upper():30s}{items[2].upper():>33s}", file=f)
+
+    # write basic information
+    _dump_integer_scalars("Number of atoms", data.natom, f)
+    _dump_integer_scalars("Number of electrons", int(data.nelec), f)
+    if data.charge is not None:
+        _dump_integer_scalars("Charge", int(data.charge), f)
+    if data.mo is not None:
+        # assign number of alpha and beta electrons
```

Comment:
Could you add a few checks for the occupation numbers?

- FCHK only supports integer occupation numbers. (It cannot be used to store natural orbitals with fractional occupation numbers.)
- FCHK assumes occupied orbitals come first, followed by virtual, which may not be the case in the IOData object.

In both cases, if the IOData is not compatible with the assumptions of the FCHK format, an exception should be raised.

In general, we need a good mechanism form "preparing" an IOData object if it can be adapted to become compatible with the assumptions of the file format. E.g. in the case of natural orbitals, the FCHK file can still contain the 1RDM. Adaptations to the IOData object are also needed in the WFN file, e.g. to decontract the basis. To keep the length of the `dump_one` and `dump_many` functions under control, it would be good to factor out such manipulations (and possibly introduce a file format API for it). @FarnazH @PaulWAyers @evohringer: we've discussed this before and I would prefer to make an issue for this. It is not something that should be handled in this PR (or stop it from merging).

Reply:
Add some lines for check if the number of occupations for orbitals alpha and beta are all integers. We need to work more in this line for the IOData object

Reply:
@tovrstra, to add the second check (occupied followed by virtual orbitals), I add a check to make sure `mo.kind` is not "generalized". Does it fully cover this check? 

I am in favor of factoring out IOData preparing before passing it to `dump_one`, and good to have an issue on that.